File: lib/youtube.ts
/**
 * YouTube Data API v3 Service
 * https://developers.google.com/youtube/v3
 *
 * For YouTube Shorts:
 * - Video must be vertical (9:16 aspect ratio)
 * - Duration must be 60 seconds or less
 * - Include #Shorts in title or description
 */

const YOUTUBE_OAUTH_BASE = "https://accounts.google.com/o/oauth2/v2/auth";
const YOUTUBE_TOKEN_BASE = "https://oauth2.googleapis.com/token";
const YOUTUBE_API_BASE = "https://www.googleapis.com/youtube/v3";
const YOUTUBE_UPLOAD_BASE = "https://www.googleapis.com/upload/youtube/v3";

// YouTube API response types
export interface YouTubeError {
  code: number;
  message: string;
  errors?: Array<{
    domain: string;
    reason: string;
    message: string;
  }>;
}

export interface YouTubeChannelInfo {
  id: string;
  title: string;
  description: string;
  customUrl?: string;
  thumbnailUrl?: string;
  subscriberCount?: number;
  videoCount?: number;
}

export interface YouTubeVideoMetadata {
  title: string;
  description: string;
  tags?: string[];
  categoryId?: string;
  privacyStatus: "public" | "private" | "unlisted";
  madeForKids?: boolean;
  selfDeclaredMadeForKids?: boolean;
}

export interface YouTubeUploadResult {
  success: boolean;
  videoId?: string;
  videoUrl?: string;
  error?: string;
  errorCode?: string;
}

export interface YouTubeTokenResult {
  success: boolean;
  accessToken?: string;
  refreshToken?: string;
  expiresIn?: number;
  scope?: string;
  tokenType?: string;
  error?: string;
  errorDescription?: string;
}

export interface YouTubeVideoStatus {
  uploadStatus: "deleted" | "failed" | "processed" | "rejected" | "uploaded";
  failureReason?: string;
  rejectionReason?: string;
  privacyStatus: string;
  publishAt?: string;
}

export interface YouTubeShortSettings {
  title?: string;
  privacyStatus?: "public" | "private" | "unlisted";
  categoryId?: string;
  madeForKids?: boolean;
  tags?: string[];
}

/**
 * Generate YouTube OAuth authorization URL
 */
export function getYouTubeAuthorizationUrl(
  clientId: string,
  redirectUri: string,
  state: string
): string {
  // Scopes for uploading videos and reading channel info
  const scopes = [
    "https://www.googleapis.com/auth/youtube.upload",
    "https://www.googleapis.com/auth/youtube.readonly",
  ].join(" ");

  const params = new URLSearchParams({
    client_id: clientId,
    redirect_uri: redirectUri,
    response_type: "code",
    scope: scopes,
    state,
    access_type: "offline", // Required for refresh token
    prompt: "consent", // Force consent screen to get refresh token
    include_granted_scopes: "true",
  });

  return `${YOUTUBE_OAUTH_BASE}?${params.toString()}`;
}

/**
 * Exchange authorization code for access token
 */
export async function exchangeYouTubeCodeForToken(
  clientId: string,
  clientSecret: string,
  code: string,
  redirectUri: string
): Promise<YouTubeTokenResult> {
  try {
    console.log("[YouTube Token Exchange] Starting token exchange...");

    const response = await fetch(YOUTUBE_TOKEN_BASE, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        client_id: clientId,
        client_secret: clientSecret,
        code,
        redirect_uri: redirectUri,
        grant_type: "authorization_code",
      }),
    });

    const result = await response.json();
    console.log("[YouTube Token Exchange] Response status:", response.status);

    if (result.error) {
      console.error("[YouTube Token Exchange] Error:", result.error, result.error_description);
      return {
        success: false,
        error: result.error,
        errorDescription: result.error_description,
      };
    }

    return {
      success: true,
      accessToken: result.access_token,
      refreshToken: result.refresh_token,
      expiresIn: result.expires_in,
      scope: result.scope,
      tokenType: result.token_type,
    };
  } catch (error) {
    console.error("[YouTube Token Exchange] Exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Refresh YouTube access token
 */
export async function refreshYouTubeAccessToken(
  clientId: string,
  clientSecret: string,
  refreshToken: string
): Promise<YouTubeTokenResult> {
  try {
    console.log("[YouTube Token Refresh] Refreshing access token...");

    const response = await fetch(YOUTUBE_TOKEN_BASE, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        client_id: clientId,
        client_secret: clientSecret,
        refresh_token: refreshToken,
        grant_type: "refresh_token",
      }),
    });

    const result = await response.json();

    if (result.error) {
      console.error("[YouTube Token Refresh] Error:", result.error);
      return {
        success: false,
        error: result.error,
        errorDescription: result.error_description,
      };
    }

    return {
      success: true,
      accessToken: result.access_token,
      expiresIn: result.expires_in,
      scope: result.scope,
      tokenType: result.token_type,
      // Note: Google doesn't return a new refresh token on refresh
    };
  } catch (error) {
    console.error("[YouTube Token Refresh] Exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Get YouTube channel information for the authenticated user
 */
export async function getYouTubeChannelInfo(
  accessToken: string
): Promise<{ success: boolean; data?: YouTubeChannelInfo; error?: string }> {
  try {
    const params = new URLSearchParams({
      part: "snippet,statistics",
      mine: "true",
    });

    const response = await fetch(`${YOUTUBE_API_BASE}/channels?${params.toString()}`, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
      },
    });

    const result = await response.json();

    if (result.error) {
      console.error("[YouTube Channel Info] Error:", result.error);
      return {
        success: false,
        error: result.error.message || "Failed to get channel info",
      };
    }

    if (!result.items || result.items.length === 0) {
      return {
        success: false,
        error: "No YouTube channel found for this account",
      };
    }

    const channel = result.items[0];
    return {
      success: true,
      data: {
        id: channel.id,
        title: channel.snippet.title,
        description: channel.snippet.description,
        customUrl: channel.snippet.customUrl,
        thumbnailUrl: channel.snippet.thumbnails?.default?.url,
        subscriberCount: parseInt(channel.statistics?.subscriberCount || "0"),
        videoCount: parseInt(channel.statistics?.videoCount || "0"),
      },
    };
  } catch (error) {
    console.error("[YouTube Channel Info] Exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Download video from URL and return as Buffer
 */
async function downloadVideoFromUrl(videoUrl: string): Promise<{
  success: boolean;
  data?: Buffer;
  size?: number;
  contentType?: string;
  error?: string;
}> {
  try {
    console.log("[YouTube] Downloading video from:", videoUrl.substring(0, 100) + (videoUrl.length > 100 ? "..." : ""));

    if (!videoUrl) {
      return {
        success: false,
        error: "Video URL is empty or undefined",
      };
    }

    const response = await fetch(videoUrl);

    if (!response.ok) {
      console.error(`[YouTube] Video download failed: ${response.status} ${response.statusText}`);
      return {
        success: false,
        error: `Failed to download video: ${response.status} ${response.statusText}`,
      };
    }

    const contentType = response.headers.get("content-type") || "video/mp4";
    const arrayBuffer = await response.arrayBuffer();
    const buffer = Buffer.from(arrayBuffer);

    console.log("[YouTube] Downloaded video:", buffer.length, "bytes, type:", contentType);

    if (buffer.length === 0) {
      return {
        success: false,
        error: "Downloaded video is empty (0 bytes)",
      };
    }

    return {
      success: true,
      data: buffer,
      size: buffer.length,
      contentType,
    };
  } catch (error) {
    console.error("[YouTube] downloadVideoFromUrl exception:", error);
    return {
      success: false,
      error: `Video download failed: ${error instanceof Error ? error.message : "Unknown error"}`,
    };
  }
}

/**
 * Initialize resumable upload session for YouTube
 * Returns upload URL for subsequent chunk uploads
 */
async function initResumableUpload(
  accessToken: string,
  metadata: YouTubeVideoMetadata,
  videoSize: number,
  contentType: string = "video/mp4"
): Promise<{ success: boolean; uploadUrl?: string; error?: string }> {
  try {
    // Prepare video resource metadata
    const videoResource = {
      snippet: {
        title: metadata.title.slice(0, 100), // YouTube max title length
        description: metadata.description.slice(0, 5000), // YouTube max description length
        tags: metadata.tags?.slice(0, 500) || [], // Max 500 tags
        categoryId: metadata.categoryId || "22", // 22 = People & Blogs (default)
      },
      status: {
        privacyStatus: metadata.privacyStatus,
        selfDeclaredMadeForKids: metadata.selfDeclaredMadeForKids ?? metadata.madeForKids ?? false,
      },
    };

    const params = new URLSearchParams({
      uploadType: "resumable",
      part: "snippet,status",
    });

    console.log("[YouTube] Initializing resumable upload:", {
      title: metadata.title.slice(0, 50),
      videoSize,
      privacyStatus: metadata.privacyStatus,
    });

    const response = await fetch(`${YOUTUBE_UPLOAD_BASE}/videos?${params.toString()}`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json; charset=UTF-8",
        "X-Upload-Content-Length": String(videoSize),
        "X-Upload-Content-Type": contentType,
      },
      body: JSON.stringify(videoResource),
    });

    if (response.status !== 200) {
      const errorBody = await response.json();
      console.error("[YouTube] Resumable upload init failed:", errorBody);
      return {
        success: false,
        error: errorBody.error?.message || `Failed to initialize upload: ${response.status}`,
      };
    }

    const uploadUrl = response.headers.get("Location");
    if (!uploadUrl) {
      return {
        success: false,
        error: "No upload URL returned from YouTube",
      };
    }

    console.log("[YouTube] Resumable upload initialized, got upload URL");
    return {
      success: true,
      uploadUrl,
    };
  } catch (error) {
    console.error("[YouTube] initResumableUpload exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Upload video data to YouTube using resumable upload
 */
async function uploadVideoData(
  uploadUrl: string,
  videoData: Buffer,
  contentType: string = "video/mp4"
): Promise<{ success: boolean; videoId?: string; error?: string }> {
  try {
    console.log("[YouTube] Uploading video data:", videoData.length, "bytes");

    const response = await fetch(uploadUrl, {
      method: "PUT",
      headers: {
        "Content-Type": contentType,
        "Content-Length": String(videoData.length),
      },
      body: new Uint8Array(videoData) as BodyInit,
    });

    const result = await response.json();

    if (!response.ok) {
      console.error("[YouTube] Video upload failed:", result);
      return {
        success: false,
        error: result.error?.message || `Upload failed: ${response.status}`,
      };
    }

    console.log("[YouTube] Video uploaded successfully, ID:", result.id);
    return {
      success: true,
      videoId: result.id,
    };
  } catch (error) {
    console.error("[YouTube] uploadVideoData exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Check video processing status
 */
export async function getYouTubeVideoStatus(
  accessToken: string,
  videoId: string
): Promise<{ success: boolean; data?: YouTubeVideoStatus; error?: string }> {
  try {
    const params = new URLSearchParams({
      part: "status,processingDetails",
      id: videoId,
    });

    const response = await fetch(`${YOUTUBE_API_BASE}/videos?${params.toString()}`, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
      },
    });

    const result = await response.json();

    if (result.error) {
      return {
        success: false,
        error: result.error.message,
      };
    }

    if (!result.items || result.items.length === 0) {
      return {
        success: false,
        error: "Video not found",
      };
    }

    const video = result.items[0];
    return {
      success: true,
      data: {
        uploadStatus: video.status.uploadStatus,
        failureReason: video.status.failureReason,
        rejectionReason: video.status.rejectionReason,
        privacyStatus: video.status.privacyStatus,
        publishAt: video.status.publishAt,
      },
    };
  } catch (error) {
    console.error("[YouTube] getVideoStatus exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Wait for video processing to complete
 */
export async function waitForYouTubeProcessing(
  accessToken: string,
  videoId: string,
  maxAttempts: number = 30,
  intervalMs: number = 10000
): Promise<YouTubeUploadResult> {
  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    console.log(`[YouTube] Checking video status, attempt ${attempt + 1}/${maxAttempts}`);

    const statusResult = await getYouTubeVideoStatus(accessToken, videoId);

    if (!statusResult.success) {
      return {
        success: false,
        videoId,
        error: statusResult.error,
      };
    }

    const status = statusResult.data!;
    console.log(`[YouTube] Video status: ${status.uploadStatus}`);

    switch (status.uploadStatus) {
      case "processed":
        console.log("[YouTube] Video processing complete!");
        return {
          success: true,
          videoId,
          videoUrl: `https://youtube.com/shorts/${videoId}`,
        };

      case "uploaded":
        // Still processing, wait and retry
        console.log(`[YouTube] Video still processing, waiting ${intervalMs}ms...`);
        await new Promise((resolve) => setTimeout(resolve, intervalMs));
        break;

      case "failed":
        console.error("[YouTube] Video processing FAILED!", status.failureReason);
        return {
          success: false,
          videoId,
          error: status.failureReason || "Video processing failed",
        };

      case "rejected":
        console.error("[YouTube] Video REJECTED!", status.rejectionReason);
        return {
          success: false,
          videoId,
          error: status.rejectionReason || "Video was rejected",
        };

      case "deleted":
        return {
          success: false,
          videoId,
          error: "Video was deleted",
        };

      default:
        console.warn(`[YouTube] Unknown status: ${status.uploadStatus}`);
        await new Promise((resolve) => setTimeout(resolve, intervalMs));
        break;
    }
  }

  // If we reach here, processing timed out but upload was successful
  // Return success since the video is uploaded, just not fully processed
  console.log("[YouTube] Processing check timed out, but video was uploaded successfully");
  return {
    success: true,
    videoId,
    videoUrl: `https://youtube.com/shorts/${videoId}`,
  };
}

/**
 * Upload a YouTube Short
 * Main function to upload a video as YouTube Short
 *
 * Requirements for Shorts:
 * - Vertical video (9:16 aspect ratio)
 * - 60 seconds or less
 * - #Shorts hashtag in title or description
 */
export async function uploadYouTubeShort(
  accessToken: string,
  videoUrl: string,
  caption: string,
  hashtags: string[],
  settings?: YouTubeShortSettings
): Promise<YouTubeUploadResult> {
  try {
    // Step 1: Download video from URL
    const downloadResult = await downloadVideoFromUrl(videoUrl);
    if (!downloadResult.success || !downloadResult.data) {
      return {
        success: false,
        error: downloadResult.error || "Failed to download video",
      };
    }

    const videoBuffer = downloadResult.data;
    const videoSize = downloadResult.size!;
    const contentType = downloadResult.contentType || "video/mp4";

    // Step 2: Prepare metadata
    // Ensure #Shorts is in the title or description for YouTube to recognize it as a Short
    const allHashtags = [...(settings?.tags || []), ...hashtags];
    const hashtagsStr = allHashtags.map((h) => (h.startsWith("#") ? h : `#${h}`)).join(" ");
    const shortsHashtag = allHashtags.some((h) => h.toLowerCase().replace("#", "") === "shorts")
      ? ""
      : "#Shorts";

    // Use settings.title if provided, otherwise use caption
    const title = (settings?.title || caption || "Untitled Short").trim().slice(0, 100);
    const description = `${caption}\n\n${shortsHashtag} ${hashtagsStr}`.trim();

    const metadata: YouTubeVideoMetadata = {
      title,
      description,
      tags: allHashtags.filter((h) => !h.startsWith("#")).concat(["Shorts"]),
      privacyStatus: settings?.privacyStatus || "public",
      categoryId: settings?.categoryId || "22", // People & Blogs
      madeForKids: settings?.madeForKids ?? false,
    };

    console.log("[YouTube Shorts] Starting upload:", {
      title: metadata.title.slice(0, 30),
      videoSize,
      privacyStatus: metadata.privacyStatus,
    });

    // Step 3: Initialize resumable upload
    const initResult = await initResumableUpload(accessToken, metadata, videoSize, contentType);
    if (!initResult.success || !initResult.uploadUrl) {
      return {
        success: false,
        error: initResult.error || "Failed to initialize upload",
      };
    }

    // Step 4: Upload video data
    const uploadResult = await uploadVideoData(initResult.uploadUrl, videoBuffer, contentType);
    if (!uploadResult.success || !uploadResult.videoId) {
      return {
        success: false,
        error: uploadResult.error || "Failed to upload video",
      };
    }

    console.log("[YouTube Shorts] Upload complete, video ID:", uploadResult.videoId);

    // Step 5: Wait for processing (optional, can return early)
    // For immediate response, we return success after upload
    // Processing status can be checked later
    const videoId = uploadResult.videoId;

    // Quick status check (don't wait for full processing)
    const statusResult = await getYouTubeVideoStatus(accessToken, videoId);
    if (statusResult.success && statusResult.data?.uploadStatus === "failed") {
      return {
        success: false,
        videoId,
        error: statusResult.data.failureReason || "Upload failed during processing",
      };
    }

    return {
      success: true,
      videoId,
      videoUrl: `https://youtube.com/shorts/${videoId}`,
    };
  } catch (error) {
    console.error("[YouTube Shorts] uploadYouTubeShort exception:", error);
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error",
    };
  }
}

/**
 * Get user info from YouTube (basic profile)
 */
export async function getYouTubeUserInfo(accessToken: string): Promise<{
  success: boolean;
  data?: {
    channelId: string;
    channelTitle: string;
    channelUrl?: string;
    thumbnailUrl?: string;
  };
  error?: string;
}> {
  const channelResult = await getYouTubeChannelInfo(accessToken);

  if (!channelResult.success || !channelResult.data) {
    return {
      success: false,
      error: channelResult.error || "Failed to get channel info",
    };
  }

  const channel = channelResult.data;
  return {
    success: true,
    data: {
      channelId: channel.id,
      channelTitle: channel.title,
      channelUrl: channel.customUrl
        ? `https://youtube.com/${channel.customUrl}`
        : `https://youtube.com/channel/${channel.id}`,
      thumbnailUrl: channel.thumbnailUrl,
    },
  };
}

File: backend/compose-engine/app/services/publishers/youtube.py
"""
YouTube Data API v3 Video Upload Service
For YouTube Shorts (<= 60 seconds vertical videos)
"""

import httpx
import asyncio
import tempfile
import os
from typing import Optional, Dict, Any, Literal
from dataclasses import dataclass, field
import logging
import json

from ...utils.s3_client import S3Client

logger = logging.getLogger(__name__)

YOUTUBE_API_BASE = "https://www.googleapis.com"
YOUTUBE_UPLOAD_URL = "https://www.googleapis.com/upload/youtube/v3/videos"


@dataclass
class YouTubeShortSettings:
    title: str = "Untitled Short"
    privacy_status: Literal["public", "private", "unlisted"] = "public"
    category_id: str = "22"  # People & Blogs
    made_for_kids: bool = False
    tags: list[str] = field(default_factory=list)


@dataclass
class YouTubePublishResult:
    success: bool
    video_id: Optional[str] = None
    video_url: Optional[str] = None
    channel_id: Optional[str] = None
    error: Optional[str] = None
    status: Optional[str] = None


class YouTubePublisher:
    """YouTube Shorts Publishing Service"""

    def __init__(self, client_id: str, client_secret: str):
        self.client_id = client_id
        self.client_secret = client_secret

    async def refresh_token(self, refresh_token: str) -> Dict[str, Any]:
        """Refresh YouTube/Google access token"""
        async with httpx.AsyncClient() as client:
            response = await client.post(
                "https://oauth2.googleapis.com/token",
                headers={"Content-Type": "application/x-www-form-urlencoded"},
                data={
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                    "grant_type": "refresh_token",
                    "refresh_token": refresh_token,
                },
            )
            result = response.json()

            if result.get("error"):
                return {
                    "success": False,
                    "error": result.get("error_description", result.get("error")),
                }

            return {
                "success": True,
                "access_token": result.get("access_token"),
                "expires_in": result.get("expires_in"),
            }

    async def _download_video(self, video_url: str) -> Dict[str, Any]:
        """Download video from URL (supports S3 URLs with credentials)"""
        try:
            logger.info(f"[YouTube] Downloading video from: {video_url[:100]}...")

            # Check if it's an S3 URL - use S3Client for authenticated download
            if ".s3." in video_url and ".amazonaws.com" in video_url:
                logger.info("[YouTube] Detected S3 URL, using S3Client for download")
                s3_client = S3Client()

                # Download to temp file
                with tempfile.NamedTemporaryFile(delete=False, suffix=".mp4") as tmp:
                    temp_path = tmp.name

                try:
                    await s3_client.download_file(video_url, temp_path)

                    with open(temp_path, "rb") as f:
                        video_data = f.read()

                    logger.info(f"[YouTube] Downloaded {len(video_data)} bytes from S3")
                    return {
                        "success": True,
                        "data": video_data,
                        "size": len(video_data),
                        "content_type": "video/mp4",
                    }
                finally:
                    if os.path.exists(temp_path):
                        os.remove(temp_path)
            else:
                # Regular HTTP download for non-S3 URLs
                async with httpx.AsyncClient(timeout=120.0) as client:
                    response = await client.get(video_url)

                    if response.status_code != 200:
                        return {
                            "success": False,
                            "error": f"Failed to download video: {response.status_code}",
                        }

                    video_data = response.content
                    content_type = response.headers.get("content-type", "video/mp4")
                    logger.info(f"[YouTube] Downloaded {len(video_data)} bytes")

                    return {
                        "success": True,
                        "data": video_data,
                        "size": len(video_data),
                        "content_type": content_type,
                    }
        except Exception as e:
            logger.error(f"[YouTube] Download error: {e}")
            return {"success": False, "error": str(e)}

    async def _get_resumable_upload_url(
        self,
        access_token: str,
        metadata: dict,
        video_size: int,
        content_type: str = "video/mp4",
    ) -> Dict[str, Any]:
        """Get resumable upload URL from YouTube"""
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{YOUTUBE_UPLOAD_URL}?uploadType=resumable&part=snippet,status",
                headers={
                    "Authorization": f"Bearer {access_token}",
                    "Content-Type": "application/json; charset=UTF-8",
                    "X-Upload-Content-Length": str(video_size),
                    "X-Upload-Content-Type": content_type,
                },
                json=metadata,
            )

            if response.status_code != 200:
                error_text = response.text
                logger.error(f"[YouTube] Init upload failed: {error_text}")
                return {
                    "success": False,
                    "error": f"Failed to initialize upload: {error_text}",
                }

            upload_url = response.headers.get("location")
            if not upload_url:
                return {"success": False, "error": "No upload URL in response"}

            return {"success": True, "upload_url": upload_url}

    async def _upload_video(
        self,
        upload_url: str,
        video_data: bytes,
        content_type: str = "video/mp4",
    ) -> Dict[str, Any]:
        """Upload video data to YouTube"""
        async with httpx.AsyncClient(timeout=300.0) as client:
            response = await client.put(
                upload_url,
                headers={
                    "Content-Type": content_type,
                    "Content-Length": str(len(video_data)),
                },
                content=video_data,
            )

            if response.status_code not in (200, 201):
                error_text = response.text
                logger.error(f"[YouTube] Upload failed: {error_text}")
                return {
                    "success": False,
                    "error": f"Upload failed: {response.status_code}",
                }

            result = response.json()
            return {
                "success": True,
                "video_id": result.get("id"),
                "status": result.get("status", {}).get("uploadStatus"),
            }

    async def _get_video_status(
        self, access_token: str, video_id: str
    ) -> Dict[str, Any]:
        """Get video processing status"""
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{YOUTUBE_API_BASE}/youtube/v3/videos",
                params={
                    "part": "status,processingDetails",
                    "id": video_id,
                },
                headers={"Authorization": f"Bearer {access_token}"},
            )

            if response.status_code != 200:
                return {"success": False, "error": "Failed to get video status"}

            result = response.json()
            items = result.get("items", [])

            if not items:
                return {"success": False, "error": "Video not found"}

            video = items[0]
            return {
                "success": True,
                "upload_status": video.get("status", {}).get("uploadStatus"),
                "privacy_status": video.get("status", {}).get("privacyStatus"),
                "processing_status": video.get("processingDetails", {}).get(
                    "processingStatus"
                ),
            }

    async def _wait_for_processing(
        self,
        access_token: str,
        video_id: str,
        max_attempts: int = 30,
        interval_sec: float = 10.0,
    ) -> YouTubePublishResult:
        """Wait for video processing to complete"""
        for attempt in range(max_attempts):
            logger.info(f"[YouTube] Processing check {attempt + 1}/{max_attempts}")
            status_result = await self._get_video_status(access_token, video_id)

            if not status_result["success"]:
                return YouTubePublishResult(
                    success=False,
                    video_id=video_id,
                    error=status_result.get("error"),
                )

            upload_status = status_result.get("upload_status")
            processing_status = status_result.get("processing_status")

            logger.info(f"[YouTube] Status: upload={upload_status}, processing={processing_status}")

            if upload_status == "processed":
                return YouTubePublishResult(
                    success=True,
                    video_id=video_id,
                    video_url=f"https://youtube.com/shorts/{video_id}",
                    status="processed",
                )

            if upload_status in ("failed", "rejected", "deleted"):
                return YouTubePublishResult(
                    success=False,
                    video_id=video_id,
                    error=f"Video {upload_status}",
                    status=upload_status,
                )

            # Still processing
            await asyncio.sleep(interval_sec)

        # Processing not complete but upload succeeded
        return YouTubePublishResult(
            success=True,
            video_id=video_id,
            video_url=f"https://youtube.com/shorts/{video_id}",
            status="processing",
        )

    async def get_channel_info(self, access_token: str) -> Dict[str, Any]:
        """Get authenticated user's channel info"""
        async with httpx.AsyncClient() as client:
            response = await client.get(
                f"{YOUTUBE_API_BASE}/youtube/v3/channels",
                params={"part": "snippet,statistics", "mine": "true"},
                headers={"Authorization": f"Bearer {access_token}"},
            )

            if response.status_code != 200:
                return {"success": False, "error": "Failed to get channel info"}

            result = response.json()
            items = result.get("items", [])

            if not items:
                return {"success": False, "error": "No channel found"}

            channel = items[0]
            snippet = channel.get("snippet", {})

            return {
                "success": True,
                "channel_id": channel.get("id"),
                "title": snippet.get("title"),
                "custom_url": snippet.get("customUrl"),
                "thumbnail_url": snippet.get("thumbnails", {})
                .get("default", {})
                .get("url"),
            }

    async def publish(
        self,
        access_token: str,
        video_url: str,
        caption: str = "",
        hashtags: list[str] = None,
        settings: Optional[YouTubeShortSettings] = None,
    ) -> YouTubePublishResult:
        """
        Upload video as YouTube Short

        Args:
            access_token: YouTube/Google OAuth access token
            video_url: URL of the video to upload
            caption: Video description
            hashtags: List of hashtags
            settings: YouTube Short settings
        """
        try:
            settings = settings or YouTubeShortSettings()
            hashtags = hashtags or []

            # Step 1: Download video
            download_result = await self._download_video(video_url)
            if not download_result["success"]:
                return YouTubePublishResult(
                    success=False, error=download_result.get("error")
                )

            video_data = download_result["data"]
            video_size = download_result["size"]
            content_type = download_result.get("content_type", "video/mp4")

            # Step 2: Prepare metadata
            # Ensure #Shorts is in the description
            all_tags = list(set(settings.tags + hashtags))
            hashtags_str = " ".join(
                h if h.startswith("#") else f"#{h}" for h in all_tags
            )
            shorts_tag = "#Shorts" if "shorts" not in [t.lower().replace("#", "") for t in all_tags] else ""

            title = settings.title[:100]
            description = f"{caption}\n\n{shorts_tag} {hashtags_str}".strip()

            metadata = {
                "snippet": {
                    "title": title,
                    "description": description,
                    "tags": [t.replace("#", "") for t in all_tags] + ["Shorts"],
                    "categoryId": settings.category_id,
                },
                "status": {
                    "privacyStatus": settings.privacy_status,
                    "selfDeclaredMadeForKids": settings.made_for_kids,
                },
            }

            logger.info(f"[YouTube] Uploading: {title[:30]}... ({video_size} bytes)")

            # Step 3: Get resumable upload URL
            init_result = await self._get_resumable_upload_url(
                access_token, metadata, video_size, content_type
            )
            if not init_result["success"]:
                return YouTubePublishResult(
                    success=False, error=init_result.get("error")
                )

            # Step 4: Upload video
            upload_result = await self._upload_video(
                init_result["upload_url"], video_data, content_type
            )
            if not upload_result["success"]:
                return YouTubePublishResult(
                    success=False, error=upload_result.get("error")
                )

            video_id = upload_result["video_id"]
            logger.info(f"[YouTube] Uploaded! Video ID: {video_id}")

            # Step 5: Wait for processing
            return await self._wait_for_processing(access_token, video_id)

        except Exception as e:
            logger.error(f"[YouTube] Publish error: {e}")
            return YouTubePublishResult(success=False, error=str(e))
